src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Formats.Cbor;
+
+namespace System.Security.Cryptography.Cose
+{
+    public class CoseHeaderMap : IEnumerable<(CoseHeaderLabel, ReadOnlyMemory<byte>)>
+    {
+        private static readonly byte[] s_emptyBstrEncoded = new byte[] { 0x40 };
+        private bool _isReadOnly;
+        public bool IsReadOnly { get => _isReadOnly; internal set => _isReadOnly = value; }
+        private readonly Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>> _headerParameters = new Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>>();
+
+        public CoseHeaderMap() : this (isReadOnly: false) { }
+
+        internal CoseHeaderMap(bool isReadOnly)
+        {
+            _isReadOnly = isReadOnly;
+        }
+
+        public bool TryGetEncodedValue(CoseHeaderLabel label, out ReadOnlyMemory<byte> encodedValue)
+            => _headerParameters.TryGetValue(label, out encodedValue);
+
+        public ReadOnlyMemory<byte> GetEncodedValue(CoseHeaderLabel label)
+            => _headerParameters[label];
```

Comment:
This will throw an exception from the dictionary if the label isn't in it.  Is that the desired behavior / exception?

Reply:
I don't know but I guess we better throw our own `InvalidOperationException`; `KeyNotFoundException` might be confusing especially since in the context of this type the dictionary keys are called "labels" (as per the RFC).